### ISSUE TITLE
WASM support?

### ISSUE THREAD

**Issue by Boscop** [jeremychone/rust-genai#11]
It would be nice to make the tokio dep optional so that it also works in wasm frontends :)

**Comment by jeremychone** [comment]
I agree that this would be cool, but it could be too much to take on for now. 

I will close this one, but mark at as "might-be-for-later"

**Comment by bachp** [comment]
I would be interested in WASM support, specifically wasip2 to be able to run genai on top of https://github.com/golemcloud/golem to build an Agent puerly in Rust.

As this was marked for mabye later almost 2 years ago, I wanted to check if this would be of interest now, before digging into what changes would be required.

**Comment by jeremychone** [comment]
@bachp This would be a good capability to add, but I worry about the impact it would have on the code. 

For example, what library it might have to pull in, and the impact it would have on the main code by default.

Do you have some insight on those?

**Comment by jeremychone** [comment]
Also, to be fully transparent, I am somewhat skeptical of WASM on the server. 

1) Native performance and capabilities for servers: After all the work on the Rust core, runtime, and ecosystem regarding performance and concurrency, introducing a middle-runtime for cloud processing has always looked counterproductive from my perspective, from a net value point of view. 

2) Running GenAI from a browser environment could be interesting, but having the keys on the client kind of goes against the best security practice of not trusting the (web browser) client with keys. So, typically, a web API to a server, and then the server handling this, would generally be the right approach for many production applications.

So, this is why regarding this capability, while it could be a popular feature, the fact that it could have a big impact on the codebase (design and implementation) and my perspective (which may be partial) on the subject, I did not spend much time thinking about this one.


Now, if this could just be an opt-in feature, with no change to the rest of the codebase, then I would be interested in considering it. 

But if it introduces changes in many core files with some sort of if/else WASM, then I will have to think hard about it.

**Comment by bachp** [comment]
@jeremychone Thanks for the feedback. I just had a quick look over to code and I think the two main things that are not compatible with WASM are reqwest and tokio.

So what I think would be needed is the following refactorings:

1. Add a trait e.g. `HttpBackend` that abstracts over reqwest and use this in the library
2. Once reqwest is optional the only remaining usage of tokio seems to be in `printer` and `bedrock` so I would disable them for wasm

These refactorings should not change the behaviour so far. The next step would than be to add wasip2 support by:

3. Add new `HttpBackend` implementation based on some wasm compatible library e.g. [golem-wasi-http](https://github.com/golemcloud/golem-wasi-http)

Regarding your concerns of wasm on the server. As I think the main benefit of using WASM there is the sandboxing capability, and in the case of golem.cloud it's also the [durable executiont](https://www.restate.dev/what-is-durable-execution) that it enables. This combined with genai would allow to build AI agents that can really focus on the logic without needing to think about infrstructure error cases.

**Comment by jeremychone** [comment]
@bachp First, thanks for the links and overview. 

At a high level, I think you are right. 

The challenge is, in effect, how intrusive it would be to the code, and how much work it is. 

Let's see if we get more interest in WASM.

**Comment by bachp** [comment]
@jeremychone I hade an AI draft what the refactoring I proopse above would look like: https://github.com/bachp/rust-genai/commit/8beab627c74dede76961786dbaf1d958f29ca3b5

Note that this doesn't implement any wasm specific functionality it shows a purely a refactoring that is removing the direct dependency from reqwest and puts it behind a abstraction trait.

**Comment by jeremychone** [comment]
@bachp First, using AI for drafting things is a good use of!

I scanned it, and I like this approach overall. Perhaps, the genai crate does not implement or know about WASM directly, but it allows other crates to use it with some sort of pluggable HTTP backend.

Now, could we make these "plugs" static, and not require dynamic dispatch?

Let's continue to brainstorm here, before real work gets started.
